scipy/optimize/tests/test_zeros.py

```diff
@@ -205,6 +206,333 @@ def test_lru_cached_individual(self, method):
+
+        kwargs = kwargs0.copy()
+        kwargs['xrtol'] = 1e-3
+        res1 = zeros._chandrupatla(self.f, *bracket, **kwargs)
+        assert_array_less(res1.xr - res1.xl, 1e-3 * np.abs(res1.x))
+        kwargs['xrtol'] = 1e-6
+        res2 = zeros._chandrupatla(self.f, *bracket, **kwargs)
+        assert_array_less(res2.xr - res2.xl, 1e-6 * np.abs(res2.x))
+        assert_array_less(res2.xr - res2.xl, res1.xr - res1.xl)
+
+        kwargs = kwargs0.copy()
+        kwargs['fatol'] = 1e-3
+        res1 = zeros._chandrupatla(self.f, *bracket, **kwargs)
+        assert_array_less(np.abs(res1.fun), 1e-3)
+        kwargs['fatol'] = 1e-6
+        res2 = zeros._chandrupatla(self.f, *bracket, **kwargs)
+        assert_array_less(np.abs(res2.fun), 1e-6)
+        assert_array_less(np.abs(res2.fun), np.abs(res1.fun))
+
+        kwargs = kwargs0.copy()
+        kwargs['frtol'] = 1e-3
+        x1, x2 = bracket
+        f0 = np.minimum(abs(self.f(x1, *args)), abs(self.f(x2, *args)))
+        res1 = zeros._chandrupatla(self.f, *bracket, **kwargs)
+        assert_array_less(np.abs(res1.fun), 1e-3*f0)
+        kwargs['frtol'] = 1e-6
+        res2 = zeros._chandrupatla(self.f, *bracket, **kwargs)
+        assert_array_less(np.abs(res2.fun), 1e-6*f0)
+        assert_array_less(np.abs(res2.fun), np.abs(res1.fun))
+
+    def test_maxiter_callback(self):
+        # Test behavior of `maxiter` parameter and `callback` interface
+        p = 0.612814
+        dist = stats.norm()
+        bracket = (-5, 5)
+        maxiter = 5
+
+        res = zeros._chandrupatla(self.f, *bracket, args=(dist, p),
+                                  maxiter=maxiter)
+        assert not np.any(res.success)
+        assert np.all(res.nfev == maxiter+2)
+        assert np.all(res.nit == maxiter)
+
+        def callback(res):
+            callback.iter += 1
+            callback.res = res
+            assert hasattr(res, 'x')
+            if callback.iter == 0:
+                # callback is called once with initial bracket
+                assert res.xl, res.xr == bracket
+            assert res.status == zeros._EINPROGRESS
+            if callback.iter == maxiter:
+                raise StopIteration
+        callback.iter = -1  # callback called once before first iteration
+        callback.res = None
+
+        res2 = zeros._chandrupatla(self.f, *bracket, args=(dist, p),
+                                   callback=callback)
+
+        # terminating with callback is identical to terminating due to maxiter
+        # (except for `status`)
+        for key in res.keys():
+            if key == 'status':
+                assert res[key] == zeros._ECONVERR
+                assert callback.res[key] == zeros._EINPROGRESS
+                assert res2[key] == zeros._ECALLBACK
+            else:
+                assert res2[key] == callback.res[key] == res[key]
+
+    @pytest.mark.parametrize('case', optimize._tstutils._CHANDRUPATLA_TESTS)
+    def test_nit_expected(self, case):
```

Comment:
I suppose we could have the same test but vectorized, calling just once the optim function?

Reply:
Yes, we could! But it's more complicated to do so, and we're already testing vectorized usage of the function elsewhere.